#! /bin/sh -e

# DP: workaround not to make libmudflap _start/_end not small data

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

From: James E Wilson <wilson@specifix.com>
Sender: gcc-patches-owner@gcc.gnu.org
To: gcc-patches@gcc.gnu.org
Subject: [committed] libmudflap _start/_end not small data
Date: Fri, 30 Sep 2005 11:17:23 -0700

I am getting gp relocation overflow errors linking in libmudflap for a
mips-elf target.  The errors are for the special _start and _end
symbols, which are used in mf-heuristics.c.  They are declared here as
simple characters, which the compiler assumes are in the small data
section.  However, these are special symbols set by the linker script,
and will never be in any small data section.

This problem is easy to fix by declaring them as unbounded arrays of
characters.  This will ensure that the compiler will never think that
they are in the small data section.  This same trick is already used in
a number of places in gcc and newlib for the same special symbols; it is
the standard solution for this problem.

This patch has been tested with a mainline x86_64-linux C and C++ build
with a libmudflap make check.  There were no regressions.  It has also
been tested with a gcc-4.0.1 mips-elf cross compiler with a number of
patches and workarounds applied.  I get 12 passes without this patch;
and 516 passes with this patch.

I have checked in the patch.
-- 
Jim Wilson, GNU Tools Support, http://www.specifix.com

2005-09-29  James E. Wilson  <wilson@specifix.com>

	* mf-heuristics.c (_end, ENTRY_POINT): Make them arrays with unknown
	bounds.

--- libmudflap/mf-heuristics.c.orig	2005-09-30 23:50:18.792570216 +0200
+++ libmudflap/mf-heuristics.c	2005-09-30 23:50:54.322168896 +0200
@@ -42,8 +42,8 @@
 #endif
 
 
-extern char _end;
-extern char ENTRY_POINT;
+extern char _end[];
+extern char ENTRY_POINT[];
 
 
 /* Run some quick validation of the given region.  
